### PR TITLE
[X86][MC] Support encoding/decoding for JMPABS

### PR DESCRIPTION
**Pull request by KanRobert**
JMPABS is a 64-bit only ISA extension, and acts as a near-direct branch with an absolute target. The 64-bit immediate operand is treated an as absolute effective address, which is subject to canonicality checks. It is in legacy map 0 and requires REX2 prefix with `REX2.M0=0` and `REX2.W=0`. All other REX2 payload bits are ignored.

blog: https://kanrobert.github.io/rfc/All-about-APX-JMPABS/

This patch
1. Extends `ExplicitVEXPrefix` to `ExplicitOpPrefix` for instrcutions requires explicit `REX2` or `EVEX`
2. Adds `ATTR_REX2` and `IC_64BIT_REX2` to put `JMPABS` , `MOV EAX, moffs32` in different tables to avoid opcode conflict 

NOTE:
1. `ExplicitREX2Prefix` can be reused by the following PUSHP/POPP instructions.
2. `ExplicitEVEXPrefix` will be used by the instructions promoted to EVEX space for EGPR.

### CODE REVIEW

File: llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h [hunk 1]
```diff
@@ -966,9 +966,12 @@ namespace X86II {
...
     NOTRACK = 1ULL << NoTrackShift,
 
-    // Force VEX encoding
-    ExplicitVEXShift = NoTrackShift + 1,
-    ExplicitVEXPrefix = 1ULL << ExplicitVEXShift
+    // Force REX2/VEX/EVEX encoding
+    ExplicitOpPrefixShift = NoTrackShift + 1,
+    ExplicitREX2Prefix = 1ULL << ExplicitOpPrefixShift,
+    ExplicitVEXPrefix = 2ULL << ExplicitOpPrefixShift,
+    ExplicitEVEXPrefix = 3ULL << ExplicitOpPrefixShift,
```
Comment (phoebewang) [inline llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h:973]: No use so far?

File: llvm/lib/Target/X86/MCTargetDesc/X86MCCodeEmitter.cpp [hunk 2]
```diff
@@ -1305,6 +1305,8 @@ PrefixKind X86MCCodeEmitter::emitREXPrefix(int MemOperand, const MCInst &MI,
       }
     }
   }
+  if (TSFlags & X86II::ExplicitREX2Prefix)
```
Comment (phoebewang) [inline llvm/lib/Target/X86/MCTargetDesc/X86MCCodeEmitter.cpp:1308]: `TSFlags & X86II::ExplicitREX2Prefix == X86II::ExplicitREX2Prefix`

File: llvm/lib/Target/X86/X86InstrControl.td [hunk 3]
```diff
@@ -188,6 +188,9 @@ let isBranch = 1, isTerminator = 1, isBarrier = 1, isIndirectBranch = 1 in {
   }
 }
 
+def JMPABS : Ii64<0xA1, RawFrm, (outs), (ins i64imm:$dst), "jmpabs\t$dst", []>,
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrControl.td:191]: JMPABS64i?

File: llvm/lib/Target/X86/X86InstrFormats.td [hunk 4]
```diff
@@ -354,7 +363,8 @@ class X86Inst<bits<8> opcod, Format f, ImmType i, dag outs, dag ins,
   string EVEX2VEXOverride = ?;
 
   bit notEVEX2VEXConvertible = 0; // Prevent EVEX->VEX conversion.
-  bit ExplicitVEXPrefix = 0; // Force the instruction to use VEX encoding.
+  ExplicitOpPrefix explicitOpPrefix = NoExplicitOpPrefix;
+  bits<2> explicitOpPrefixBits = explicitOpPrefix.Value;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrFormats.td:367]: Why need to define both?

File: llvm/utils/TableGen/X86RecognizableInstr.h [hunk 5]
```diff
@@ -155,6 +155,8 @@ namespace X86Local {
   enum {
     AdSize16 = 1, AdSize32 = 2, AdSize64 = 3
   };
+
+  enum { ExplicitREX2 = 1 };
```
Comment (phoebewang) [inline llvm/utils/TableGen/X86RecognizableInstr.h:159]: No need enum for one entry.

**Comment by phoebewang** [comment]
> ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> 
> You can test this locally with the following command:
> View the diff from clang-format here.

Update the format.

File: llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h (hunk 1 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h:973]: Yes. As I said in the description. "ExplicitEVEXPrefix will be used by the instructions promoted to EVEX space for EGPR."

Comment (phoebewang) [reply llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h:973]: It seems saying in description is less efficient than saying around the code :)

File: llvm/lib/Target/X86/X86InstrFormats.td (hunk 4 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrFormats.td:367]: We need `bits<2>` member b/c in X86RecognizableInstr.cpp, `byteFromRec` requires the `bits` format.

We need `ExplicitOpPrefix` member b/c `explicitOpPrefix = ExplicitVEX` in X86InstrSSE.td.

Similarly, we have both
```
  OperandSize OpSize = OpSizeFixed; // Does this instruction's encoding change
                                    // based on operand size of the mode?
  bits<2> OpSizeBits = OpSize.Value;
  AddressSize AdSize = AdSizeX; // Does this instruction's encoding change
                                // based on address size of the mode?
  bits<2> AdSizeBits = AdSize.Value;
```
in this file.

File: llvm/utils/TableGen/X86RecognizableInstr.h (hunk 5 above)
Comment (KanRobert) [reply llvm/utils/TableGen/X86RecognizableInstr.h:159]: If no this enum, then in X86RecognizableInstr.cpp, the code would be
```
  ExplicitREX2Prefix =
      byteFromRec(Rec, "explicitOpPrefixBits") == 1;
```
which was hard to read.

File: llvm/lib/Target/X86/MCTargetDesc/X86MCCodeEmitter.cpp (hunk 2 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/MCTargetDesc/X86MCCodeEmitter.cpp]: Done

File: llvm/lib/Target/X86/X86InstrControl.td (hunk 3 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrControl.td]: Done.

File: llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h (hunk 1 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h:973]: Sure, let me add it in the code.

Comment (KanRobert) [reply llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h:977]: Done

**Comment by KanRobert** [comment]
Ping @phoebewang ?

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**